{doc} Add subprocess security guidance doc (#29681)
Azure/azure-cli: doc/cli_subprocess_guidelines.md

# Azure CLI Subprocess Guidelines

> In certain cli modules, there are scenarios that need to call a subsystem to run commands outside cli, like getting kubectl info in aks, or deployment setup in mysql using `git` and `gh`, through python built-in [subprocess](https://docs.python.org/3/library/subprocess.html) module. Despite its simplicity and versatility, ensuring the security of applying it can be challenging in different platforms under different circumstance, and it is error-prone for developers to neglect security best practices during development.

Reviewer: We usually use upper case "CLI".

> In certain cli modules, there are scenarios that need to call a subsystem to run commands outside cli, like getting kubectl info in aks, or deployment setup in mysql using `git` and `gh`, through python built-in [subprocess](https://docs.python.org/3/library/subprocess.html) module. Despite its simplicity and versatility, ensuring the security of applying it can be challenging in different platforms under different circumstance, and it is error-prone for developers to neglect security best practices during development.

Reviewer: Why using subsystem instead of subprocess? I don't think "subsystem" stands for the scenario this document is talking about.

https://docs.python.org/3/library/subprocess.html doesn't mention the term "subsystem" at all.

Reviewer: Will it be clearer to use "subprocess" instead of "subsystem"? I don't think "subsystem" stands for the scenario this document is talking about.

https://docs.python.org/3/library/subprocess.html doesn't mention the term "subsystem" at all.

> In certain cli modules, there are scenarios that need to call a subsystem to run commands outside cli, like getting kubectl info in aks, or deployment setup in mysql using `git` and `gh`, through python built-in [subprocess](https://docs.python.org/3/library/subprocess.html) module. Despite its simplicity and versatility, ensuring the security of applying it can be challenging in different platforms under different circumstance, and it is error-prone for developers to neglect security best practices during development.

Reviewer: Will it be clearer to use "subprocess" instead of "subsystem"? I don't think "subsystem" stands for the scenario this document is talking about.

https://docs.python.org/3/library/subprocess.html doesn't mention the term "subsystem" at all.

Author: `subsystem` is taken as a synonym or a phase in the meaning nearby to `subprocess` here.

## Insecure Use Of Subsystem Commands Under Subprocess Module

> ### Dynamic Command Construction Under Risk

Reviewer: No need to capitalize the first letter of each word.

```suggestion
### Dynamic command construction under risk
```

Assume a script that needs to read in `user_input` from users to execute a `git` related command and runs it through subprocess using shell. In regular cases, users would add a valid git command, like `git --version` or something else, but if undesired input gets logged into this system, like `--version;echo aa`, the appended `;echo aa` would be executed by the subsystem too, like below: 

```commandline
import subprocess
user_input = input("input a git command to run: ")
cmd = f"git {user_input}"
subprocess.run(cmd, shell=True)
```

```commandline
input a git command to run: --version;echo aa
git version 2.34.1
aa
```

> This is a simple example for demonstrating the side effects in python's subsystem improper usage. And it's common for cli developers to build and execute commands dynamically from users' input in a more complicated way. When constructing and executing commands in subprocess through `shell=True`, it exposes a big security vulnerability to potential malicious users outside. 

Reviewer: subsystem -> subprocess?

https://docs.python.org/3/library/subprocess.html#older-high-level-api doesn't mention the term "subsystem" at all.

> This is a simple example for demonstrating the side effects in python's subsystem improper usage. And it's common for cli developers to build and execute commands dynamically from users' input in a more complicated way. When constructing and executing commands in subprocess through `shell=True`, it exposes a big security vulnerability to potential malicious users outside. 

Reviewer: subsystem -> subprocess?

https://docs.python.org/3/library/subprocess.html doesn't mention the term "subsystem" at all.

## Mitigating Security Vulnerability When Calling Subsystem Commands

There are several aspects of security practices that developers need to have in mindset to safeguard their cli modules from command injection attacks.

### Cli Centralized Subsystem Executing

Azure cli provides a centralized function `run_cmd` adapted from official `subprocess.run`, with necessary argument covered and illegal input blocking enforced. 

What developers need to do is:

> 1) `from azure.cli.core.util import run_cmd`
> 2) replace `subprocess.run` (or `Popen` or `check_call` or `check_output` or `call`) with `run_cmd`.
> 3) construct cmd args as array like: [executable, arg0, arg1, arg2, ...]

Reviewer: ```suggestion
1. `from azure.cli.core.util import run_cmd`
2. replace `subprocess.run` (or `Popen` or `check_call` or `check_output` or `call`) with `run_cmd`.
3. construct cmd args as array like: `[executable, arg0, arg1, arg2, ...]`
```

`run_cmd` will add necessary argument type checks and process the input and output the same way as `subprocess.run`, and block potential risks from commands constructed from user input.
Below is an example for `run_cmd` use case:

> ```commandline

Reviewer: ```suggestion
```py
```

# code before:
cmd = f"git commit -m {user_message}"
import subprocess
output = subprocess.run(cmd, shell=True, check=False, stderr=subprocess.PIPE, stdout=subprocess.PIPE)
if output.returncode != 0:
      raise CLIInternalError('Command execution failed, command is: '{}', error message is: {}'.format(cmd, output.stderr))
return output.stdout
```

If `run_cmd` is applied, it would be like:
```commandline
# code after:
cmd = ["git","commit", "-m", user_message]
import subprocess
from azure.cli.core.util import run_cmd
output = run_cmd(cmd, check=False, capture_output=True)
if output.returncode != 0:
      raise CLIInternalError('Command execution failed, command is: '{}', error message is: {}'.format(cmd, output.stderr))
return output.stdout
```

All various kinds of `subprocess` Popen calling use cases can be easily adjusted into `run_cmd` with security risks processed and eliminated in this centralized function.

Besides that, users might need to know some parts of the accessibility in both `run_cmd` and `subprocess`

> 1) when calling shell built-in cmds, like `dir` or `echo`, using `shell=True` **in windows platform**, `subprocess` implicitly uses `cmd.exe`, while `run_cmd` asks developers to provide the `cmd.exe` as executable file specifically in the arg list's first item, like `["cmd.exe", "/c", "echo", "abc"]`

Reviewer: ```suggestion
1) When calling shell built-in cmds, like `dir` or `echo`, using `shell=True` **in windows platform**, `subprocess` implicitly uses `cmd.exe`, while `run_cmd` asks developers to provide the `cmd.exe` as executable file specifically in the arg list's first item, like `["cmd.exe", "/c", "echo", "abc"]`
```

2) if developers want to find an easy way to split their current cmd string into list, **for unix-like platforms**, developers can apply [`shlex.split`](https://docs.python.org/3/library/shlex.html#shlex.split) for quick access. But a prepared cmd statement is still more recommended (for more info about prepared cmd statement, please read below sections).

> 3) it might be not that obvious to find target command's executable file **in windows platform**, a tool developer can use is `shutil.which` that gives the executable file path in windows system, like `shutil.which(git)`. The cmd `git --version` can be adjusted as `[shutil.which(git), "--version"]`. Please provide the corresponding executable path in target platforms.

Reviewer: Could you share more information on this? I think `subprocess` will look for the executable on `PATH`.

Author: Yes, subprocess searches executable under `PATH`, as the same way as `shutil.which`. 

In windows platform, cmd-running can be `cmd arg0 arg2..` or `cmd.exe arg0 arg1...`. Just want to give a hint to users how to check the absolute path of corresponding executable file in case they are confused with it. 

Doc adjusted.

```commandline
import re
def sanitize_input(user_input):
    return re.sub(r'[^a-zA-Z0-9]', '', user_input)
```
In a real cli module scenario, developers better use their corresponding sanitization pattern for reducing the chaining commands risks.

#### Use parameterized cmd statement

A parameterized cmd statement is a way to structure a command starting from defining the command structure first and then provide the parameters that should be inserted into the command separately. 
In this way, the system will treat the first item as the executable file and the left item as args of that executable application.
```
# instead of
cmd = f"git {user_input}"
# using 
cmd = ["git", user_input]
```

#### Avoid use `shell=True` with user_inputs

When using subprocess module, avoid `shell=True` argument when it comes with cmd containing user inputs. When python executes a command with `shell=True`, the system shell will interpret any string passed in and if this string contains user input without proper sanitization, an attacker can inject malicious commands, as demonstrated in the very beginning of this doc.

## Summary

> Ensuring the safety of cli from command injection under subprocess calling requires an in-depth understanding of these vulnerabilities and also proactive measures to counteract potential exploits. Cli developers can either apply the three security practices, if applicable, when using builtin `subprocess`, or use a more centralized func `run_cmd` cli provided, to safeguard cli modules from command injection attack and for future more accessible security enforcements.

Reviewer 2: Recommend them to use the centralized function `run_cmd` first, if it cannot meet their requirement, then they can write their own function with applying the three security practices.

> Ensuring the safety of cli from command injection under subprocess calling requires an in-depth understanding of these vulnerabilities and also proactive measures to counteract potential exploits. Cli developers can either apply the three security practices, if applicable, when using builtin `subprocess`, or use a more centralized func `run_cmd` cli provided, to safeguard cli modules from command injection attack and for future more accessible security enforcements.

Reviewer 2: Avoid to use abbreviation `func`, use `function` instead which is more straightforward.

## Insecure Use Of Commands Under Subprocess Module

### Dynamic command construction under risk

Assume a script that needs to read in `user_input` from users to execute a `git` related command and runs it through subprocess using shell. In regular cases, users would add a valid git command, like `git --version` or something else, but if undesired input gets logged into this system, like `--version;echo aa`, the appended `;echo aa` would be executed by `subprocess` too, like below: 

```python
import subprocess
user_input = input("input a git command to run: ")
cmd = f"git {user_input}"
subprocess.run(cmd, shell=True)
```

```python
input a git command to run: --version;echo aa
git version 2.34.1
aa

> ```

Reviewer: This is not Python code.

Author: right, it is a mock for the console input and output of running python code above

2. if developers want to find an easy way to split their current cmd string into list, **for unix-like platforms**, developers can apply [`shlex.split`](https://docs.python.org/3/library/shlex.html#shlex.split) for quick access. But a prepared cmd statement is still more recommended (for more info about prepared cmd statement, please read below sections). 
3. if developers want to locate the target command's executable file, a tool developers can use is `shutil.which` that gives the full executable file path in system, like `shutil.which(git)` returns the full `git.exe` path in windows platform `C:\\Program Files\\Git\\cmd\\git.EXE`. 
4. if the target cmd is `az`-related, like `az group show --name xxxx`, please use internal corresponding function call `cli_ctx.invoke(["az", "group", "show", "--name", "xxx"])` to get the target information.

### Best practices in subprocess use cases

> The following sections discuss some secure coding conventions that, when implemented, can help protect CLI applications from command injection vulnerabilities when calling subsystems through `subprocess`.

Reviewer: ```suggestion
The following sections discuss some secure coding conventions that, when implemented, can help protect CLI applications from command injection vulnerabilities when using `subprocess` module.
```

```python
import re
def sanitize_input(user_input):
    return re.sub(r'[^a-zA-Z0-9]', '', user_input)
```
In a real CLI module scenario, developers better use their corresponding sanitization pattern for reducing the chaining commands risks.

#### Use parameterized cmd statement

A parameterized cmd statement is a way to structure a command starting from defining the command structure first and then provide the parameters that should be inserted into the command separately. 
In this way, the system will treat the first item as the executable file and the left item as args of that executable application.

```python
# instead of
cmd = f"git {user_input}"
# using 
cmd = ["git", user_input]
```

#### Avoid use `shell=True` with user_inputs

When using subprocess module, avoid `shell=True` argument when it comes with cmd containing user inputs. When python executes a command with `shell=True`, the system shell will interpret any string passed in and if this string contains user input without proper sanitization, an attacker can inject malicious commands, as demonstrated in the very beginning of this doc.

## Summary

> Ensuring the safety of Azure CLI from command injection under subprocess calling requires an in-depth understanding of these vulnerabilities and also proactive measures to counteract potential exploits. CLI developers can apply the three security practices, if applicable, when using builtin `subprocess`, but it's recommended to use the centralized function `run_cmd` CLI provided, to safeguard CLI modules from command injection attack and for future more accessible security enforcements.

Reviewer 3: Should we have a more prescriptive approach and require CLI developers to use the `run_cmd` for any subprocess operation to meet our security standards?

Author: We can add a checking rule on those unsupervised subprocess call and give cli core's `run_cmd` as the substitution, with this doc provided as assistance.
